[JENKINS-24278] Node monitor for free inodes

batmat opened the pull request:
Adding a node monitor to get the used inodes usage percentage.

The code is actually calling the `df` command, since after a quick search I didn't find a Java lib to handle that. The calls to `df` have had to be actually more complex than I would have liked. Here's [why](https://github.com/batmat/inodes-monitor-plugin/blob/master/SAMPLE-DF-OUTPUTS.asciidoc) (btw, that file may be useful added to that PR for reference too, let me know what you think).

Created at [first a new plugin](https://github.com/batmat/inodes-monitor-plugin), but after [discussing](https://issues.jenkins-ci.org/browse/JENKINS-24278) with @oleg-nenashev we thought it might be a useful addition to core (as inodes monitoring is actually as important as disk space monitoring).

Also added a screenshot of the result of this code (using the plugin above) running on our production server.
![inodesmonitor-prod-screenshot](https://cloud.githubusercontent.com/assets/223853/12249957/8fd3cf8c-b8c3-11e5-9b8b-39e40e87c280.png)

Ping @jenkinsci/code-reviewers

batmat committed:
[JENKINS-24278] Node monitor for free inodes

Adding a node monitor using the 'df' command to get the used inodes
percentage.

oleg-nenashev commented on core/src/main/java/hudson/node_monitors/DfRunner.java:
```diff
@@ -0,0 +1,132 @@
+		IMPLEMENTATIONS.put("aix", new AixDfCommand());
+	}
+
+	public String getUsedInodesPercentage() {
```
Makes sense to create Javadoc and mention the method must be launched on master

batmat replied:
I'm OK w/ documenting, but not sure I understand your sentence. That code is called from https://github.com/jenkinsci/jenkins/pull/1974/files#diff-0b9bd711efdb6ae3c27ba783df639657R159 so it's gonna run from all slaves/nodes, isn't it? Not just from master?
Thanks

oleg-nenashev replied:
The implementation is right in your code.
My proposal was to document the behavior in order to avoid misusages of the new public API by other core/plugin developers

batmat replied:
OK, but btw I also put the DfRunner as package visible to prevent external usage. Gonna enrich javadoc too.

oleg-nenashev commented on core/src/main/java/hudson/node_monitors/DfRunner.java:
```diff
@@ -0,0 +1,132 @@
+	}
+
+	/**
+	 * Tries to run df anyway. Fallback. Will return N/A anyway if an error occurs. Or should we just return N/A directly?
```
I would just return N/A
Ideally you could add a popup or whatever else to the table's cell. This popup could mention that Jenkins currently does not support "df"

batmat replied:
Done, added a hint saying it's an "unsupported platform"

oleg-nenashev commented:
Are there any other use-cases for DfRunner? We calculate DiskUsages in the different way, so maybe makes sense to replace <code>DfX</code> by <code>INodeX</code>

olivergondza commented on core/src/main/java/hudson/node_monitors/DfRunner.java:
```diff
@@ -0,0 +1,132 @@
+import java.util.logging.Logger;
+
+class DfRunner {
+	private static final Logger LOGGER = Logger.getLogger(DfRunner.class.getSimpleName());
```
Logger should use full class name. Also, I do not like the class name (`InodeMonitor`).

batmat replied:
Thinking about it, it may indeed be incorrectly named. 
Would you prefer something like InodesUsageMonitor? Is that your point?

olivergondza commented on core/src/main/java/hudson/node_monitors/InodesMonitor.java:
```diff
@@ -0,0 +1,162 @@
+/**
+ * Checks the used percentage of inodes on the FS (Linux/Unix only).
+ *
+ * Uses the <code>df</code> system command.
```
This is an implementation detail of particular system specific implementation.

batmat replied:
Absolutely. Like for the @oleg-nenashev's comment above, gonna remove it and "hide" it a bit more to ease potential future implementation change for the retrieval.

olivergondza commented on core/src/main/java/hudson/node_monitors/InodesMonitor.java:
```diff
@@ -0,0 +1,162 @@
+ */
+public class InodesMonitor extends NodeMonitor {
+
+	private static final Logger LOGGER = Logger.getLogger(InodesMonitor.class.getSimpleName());
```
`getName()`

olivergondza commented on core/src/main/java/hudson/node_monitors/InodesMonitor.java:
```diff
@@ -0,0 +1,162 @@
+		}
+		catch (ParseException e) {
+			// Shouldn't happen since received value is the one already provided by internal GetInodesUseInPercent
+			throw new IllegalStateException("WTF? Can't parse " + currentValueStr + " as integer percentage", e);
```
However unlikely, the message should be reworded.

batmat replied:
Sure, I missed that one before pushing. Thanks

olivergondza commented on core/src/main/java/hudson/node_monitors/InodesMonitor.java:
```diff
@@ -0,0 +1,162 @@
+		return Jenkins.getInstance().hasPermission(Jenkins.ADMINISTER) ? super.getColumnCaption() : null;
+	}
+
+	@Extension
```
Annotating field is deprecated, annotate class itself and remove the field.

batmat replied:
Was unaware of that. Used the existing NodeMonitor impls to get mine running but they are certainly too old to have got that applied. Makes sense. Will do.

olivergondza commented on core/src/main/resources/hudson/node_monitors/InodesMonitor/config.jelly:
```diff
@@ -0,0 +1,29 @@
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:f="/lib/form">
+  <f:entry title="${%Inodes Usage Percentage Threshold}" field="inodesPercentThreshold">
+    <f:textbox default="96%"/>
```
`DEFAULT_OFFLINE_THRESHOLD = "95%"`

olivergondza commented:
Right now, there are disk space monitors for workspace and /tmp. This PR add inode monitoring for users home. IMO we should monitor both space and inodes on all three of the locations. Unfortunately, that would create 6 monitor columns and clutter the UI a lot. How about merging the inode monitoring into particular disk space monitor (to merge the columns and number of remoting calls)?

jtnord commented on core/src/main/java/hudson/node_monitors/InodesMonitor.java:
```diff
@@ -0,0 +1,162 @@
+
+	private static final Logger LOGGER = Logger.getLogger(InodesMonitor.class.getSimpleName());
+
+	private static final String DEFAULT_OFFLINE_THRESHOLD = "95%";
```
From experience this metic (as a percentage) is not very useful except for small systems.

When you have a lot of small files (think javadoc) archived for a large amount of time - the number of inodes is HUUGEE.  You can have millions free, yet that can be only 2% (and a million will last for say several months - and should not take a node offline) - even 1% is a little too corse.

batmat replied:
At first, I though about using the number value, and then I realized this may be simpler to use percentage, as I also wanted to enable the auto-offline feature too. And as my current code doesn't support commas, this may indeed be an issue with big configs.

jtnord commented:
Personally I would say the opposite of @oleg-nenashev and that this should be a plugin - especially so if you consider percentages for large systems do not scale well.  There are plenty of plugins that are useful by default - but that does not mean the code should be moved into core.  Where possible we are trying to get code out of core and into plugins, partly so fixes can be delivered in a matter of hours (for a plugin release) vs potentially weeks for a new LTS.  Given that I would be :-1: on including this in core - but :8ball: for a plugin.

batmat committed:
LOGGER: getName() instead of getSimpleName()

batmat committed:
Useless imports

batmat committed:
DfRunner renamed InodesUsageGetter

batmat committed:
Document this is gonna run where launched

I.e. if you for example need to run it on another node, you'll have to
wrap it in a Callable, etc.

batmat committed:
Simply return "N/A..." as a default

batmat committed:
Impl details: removed.

batmat committed:
Removed swear word

batmat committed:
Remove DESCRIPTOR field deprecated way

@olivergondza informed me this way was deprecated. Thanks for the
heads-up: https://github.com/jenkinsci/jenkins/pull/1974/files#r49419386

batmat committed:
95% default value!

batmat commented:
@olivergondza @oleg-nenashev I just pushed a commit that should hopefully address many of your remarks.

For @olivergondza's point about monitor the many available FS just like DiskSpaceMonitor, you are certainly right. I will have to dig in the code for that one. But from looking at the existing impl, I was under the impression a NodeMonitor was designed to take only one threshold value when an associated configure.groovy view was used. 

IIUC, what you'd have in mind is to just return for the column title &value both informations separated by a slash or something?

@jtnord As described, that's indeed what I did initially. But my point (and maybe @oleg-nenashev's too IIUC) is that that inodes usage monitoring should be available along disk space monitoring, since it's just as important.

So, now, if you agree with that statement: what about we reverse the strategy (maybe that's what you implied in your comment btw), and as this is what generally is trying to be done as of lately: 

do you think it would make sense to extract the existing monitoring classes in a new dedicated _node monitoring plugin_? 

That would leave the core yet a bit leaner, and give possibly more room for that (new) plugin to be improved and upgraded by users in a way that doesn't force them to upgrade the whole installation for that. 

WDYT?

batmat commented:
Side note: pushed many commits, almost one for each remark. If this PR finally gets in a state where people want to get it merged, I'll squash the whole thing before (or file a new dedicated PR depending on reviewers' preferences).

jtnord commented:
@batmat - yes I was trying to say even if important I think this should be in a plugin.
(similar to ssh slaves - which are more vital than monitoring for many installations :-) )

batmat commented:
Absolutely. In the end, I understand and agree to try and make the core smaller and more dedicated.

In that case, how do we go forward?

I suppose I could just send a PROPOSAL to the dev ML, somehow like last Jesse's email about splitting harness, to get people's feeling about it?

WDYT @olivergondza & @oleg-nenashev ?

olivergondza commented:
I agree on splitting it into the plugin.

KostyaSha commented:
:-1: for having it in core. Current NodeMonitor API is ugly, afair it impossible to disable single checks, impossible to do single checks from plugins, it's async and it good that nobody touched it for a long time. Would be perfect at all to have only generic classes in core and move all kinds of checks into separate plugin, so people may implement any kind of checks on slaves.

:-1: for wrapping binaries, as soon as any binary will change text output (i.e. like ifconfig did) all implementation would be screwed.
As i already said to @oleg-nenashev jenkins has JNA classes that is native java support.

batmat commented:
For reference, split proposal [sent yesterday to dev ML](https://groups.google.com/forum/#!topic/jenkinsci-dev/WT8VLf5lZx4).

batmat commented:
As for not wrapping binary call in favor of JNA, I don't know. Sure the output could change, even if unlikely. Anyway I'll re-dig later on that one.

Gonna close that PR now as the next step is more heading towards dedicating a new plugin for it.

Thank you all for the review & feedback.

KostyaSha commented:
@batmat JNA is also very specific, but it more clear way of getting info and it already used. With JNA you can have native calls to glibc.

batmat commented:
Yes, I think I get the idea. Never used JNA, but used JNI a bit more than 10 years ago. From what I understand JNA is basically JNI without a substantial of the JNI's enormous PITA.
